Add instructions re: SXGs using Web Packager (#4681)
GoogleChrome/web.dev: src/site/content/en/fast/signed-exchanges-webpackager/index.md

---
layout: post
title: How to set up Signed Exchanges using Web Packager
subhead: |
  Learn how to serve signed exchanges (SXGs) using Web Packager.
authors:
  - katiehempenius
date: 2020-02-16
hero: hero.jpg
alt: Abstract photo
description: |
  Learn how to serve signed exchanges (SXGs) using Web Packager.
tags:
  - blog
  - performance
---

> The following instructions explain how to set up Signed Exchanges (SXGs) using
> Web Packager. Instructions are included for both self-signed and
> `CanSignHttpExchanges` certificates.

Reviewer: ```suggestion
The following instructions explain how to set up [Signed Exchanges (SXGs)](/signed-exchanges) using
[Web Packager](https://github.com/google/webpackager). Instructions are included for both self-signed and
`CanSignHttpExchanges` certificates.
```

    ```

    A certificate signing request is a block of encoded text that conveys the
    information necessary to request a certificate from a [certificate authority(CA)](https://en.wikipedia.org/wiki/Certificate_authority). Although you will not be requesting a certificate from a
    CA, it is still necessary to create a certificate signing request.

    The command above creates a certificate signing request for an organization
    named `Web Packager Demo` that has the [common
    name](https://knowledge.digicert.com/solution/SO7239.html) `example.com`. The
    common name should be the fully qualified domain name of the site that contains
    the content that you want to package as SXG.

    In a production SXG setup, this would be a site that you own. However, in a
    testing environment like the one described in these instructions, it can be any
    site.

3. Create a certificate that has the `CanSignHttpExchanges` extension.

    ```shell
    openssl x509 -req -days 90 -in cert.csr -signkey priv.key -out cert.pem -extfile &lt;(echo -e "1.3.6.1.4.1.11129.2.1.22 = ASN1:NULL\nsubjectAltName=DNS:example.com")
    ```

>     This command uses the private key and CSR created in steps 1 and 2 to create the
>     certificate file `cert.pem`. The `-extfile` flag associates the certificate with

Reviewer: ```suggestion
    This command uses the private key and the CSR created in steps 1 and 2 to create the
    certificate file `cert.pem`. The `-extfile` flag associates the certificate with
```

3. Open `webpkgserver.toml` with an editor of your choice and make the following
   changes:
    *   Change the line `#AllowTestCert = false` to `AllowTestCert = true`.
    *   Change the line `PEMFile = 'path/to/your.pem'` to reflect the path to the
        PEM certificate, `cert.pem`, that you created. Do not change the line
        mentioning `TLS.PEMFile`—this is a different configuration option.
    *   Change the line `KeyFile = 'priv.key'` to reflect the path of the private
        key, `priv.key`, that you created. Do not change the line mentioning
        `TLS.KeyFile`—this is a different configuration option.
    *   Change the line `Domain = 'example.org'` to reflect the domain that you
        created a certificate for. If you have followed the instructions in this
        article verbatim, this should be changed to `example.com`. `webpkgserver`
        will only fetch content from the domain indicated by `webpkgserver.toml`. If
        you try to fetch pages from a different domain without updating
        `webpkgserver.toml`, the `webpkgserver` logs will show the error message
        `URL doesn't match the fetch targets`.

>     *   (Optional) To preload subresources, change the line `#PreloadCSS
>         = false` to `PreloadCSS = true`. In addition, change the line `#PreloadJS =
>         false` to `PreloadJS = true`. More information about subresource
>         substitution can be found
>         [here](https://github.com/WICG/webpackage/blob/master/explainers/signed-exchange-subresource-substitution.md).

Reviewer: ```suggestion
    *   (Optional) To preload subresources, change the line `#PreloadCSS
        = false` to `PreloadCSS = true`. In addition, change the line `#PreloadJS =
        false` to `PreloadJS = true`. For more information about subresource
        substitution, check out
        [the explainer](https://github.com/WICG/webpackage/blob/master/explainers/signed-exchange-subresource-substitution.md).
```

  to fetch certificate`. The browser displays this error when it is unable to
  load a certificate from the **Certificate URL** indicated in the signature. The
  next section explains how to fix this error by uploading a certificate.

  Without the certificate, the browser is unable to authenticate the SXG and it
  falls back to loading the resource without using SXG. This is why there is an
  additional request to `example.com` listed in the **Network** panel.

  {% Img
    src="image/j2RDdG43oidUy6AL6LovThjeX9c2/bU5xNjhvVKr9eSg4RqEZ.png",
    alt="Screenshot of the DevTools Network tab loading 'example.com' 
    without using SXG.",
    width="800",
    height="192"
  %}

### Upload the self-signed certificate

  To establish the authenticity of a SXG, the browser must be able to load the
  certificate that was used to sign the SXG from the **Certificate URL** indicated in
  the signature. If the browser is unable to load this certificate, it will
  request that the content be delivered without using SXG.

  These instructions explain how to serve a certificate from an existing HTTPS   environment. 

  {% Aside %}

>   `webpkgserver` can be configured to use a locally-hosted
>   certificate. For information on this configuration option,
>   refer to the `CertURLBase` option in `webpkgserver.toml`. This 
>   [article](https://web.dev/how-to-use-local-https/) explains how to 
>   use HTTPS for local development.

Reviewer: ```suggestion
  `webpkgserver` can be configured to use a locally-hosted
  certificate. For information on this configuration option,
  refer to the `CertURLBase` option in `webpkgserver.toml`. 
  Learn more about [how to use HTTPS for local development](https://web.dev/how-to-use-local-https/).
```

  {% endAside %}

#### Prerequisites

  The instructions in this section assume that you have the ability to upload a
  certificate to an existing HTTPS site. In addition, you should be comfortable
  adjusting the server configuration of this site.

#### Instructions

1. In DevTools, locate the **Certificate URL** indicated in the
   **Signature** of the SXG. Copy the hash that is located at the end of this string.

  This hash is an identifier that corresponds to the certificate. If you were to
  regenerate the SXG using a different certificate, the **Certificate URL**
  listed in the **Signature** would be different.

  {% Img
    src="image/j2RDdG43oidUy6AL6LovThjeX9c2/K1caujHPlyCURA1pxYXz.png",
    alt="Screenshot showing the hash that corresponds to the certificate.",     width="800",
    height="229"
  %}

> 2. Create a copy of `cert.pem`. The filename of the new version should match the

Reviewer: Seems like the image here breaks the continuity of the numbered list, the steps are rendered as 1, 1, 2, etc

possible to authenticate the origin of a resource independently of how it was delivered. 
The following instructions explain how to set up Signed Exchanges using
[Web Packager](https://github.com/google/webpackager). Instructions are included for 
both self-signed and `CanSignHttpExchanges` certificates.

## Serve SXGs using a self-signed certificate

Using a self-signed certificate to serve SXGs is primarily used for
demonstration and testing purposes. SXGs signed with a self-signed certificate
will generate error messages in the browser when used outside of testing
environments and should not be served to crawlers.

### Prerequisites

To follow these instructions you will need to have
[openssl](https://github.com/openssl/openssl#build-and-install) and
[Go](https://golang.org/doc/install) installed in your development environment.
You will also need an existing HTTPS site.

### Architecture

These instructions use the following architecture to serve SXGs:

{% Img 
  src="image/j2RDdG43oidUy6AL6LovThjeX9c2/cyV0pNmlTB4PTCc5YQe3.png",
  alt="Signed Exchanges are loaded from a Web Packager instance running 
  on localhost:8080",
  width="800",
  height="436"
%}

> Following these instructions verbatim will setup a `webpackager` instance that

Reviewer 2: `webpackager` HTTP server instance

    `priv.key` and `cert.pem` files in the next section.

### Setup the Web Packager server for testing

#### Prerequisites

1. Install [Web Packager](https://github.com/google/webpackager).

    ```shell
    git clone https://github.com/google/webpackager.git
    ```

2. Build `webpkgserver`.

    ```shell
    cd webpackager/cmd/webpkgserver
    go build .
    ```

    `webpkgserver` is a specific binary within the Web Packager project.

3. Verify that `webpkgserver` has been installed correctly.

    ```shell
    webpkgserver --help
    ```

    This command should return information about the usage of `webpkgserver`. If
    this does not work, a good first troubleshooting step is to verify that your
    [GOPATH](https://golang.org/doc/gopath_code.html#GOPATH) is configured
    correctly.

#### Instructions

1. Navigate to the `webpackager` directory (you might already be in this
   directory).

    ```shell
    cd /path/to/cmd/webpkgserver
    ```

2. Create a `webpkgsever.toml` file by copying the example.

    ```shell
    cp ./webpkgserver.example.toml ./webpkgserver.toml
    ```

    This file contains the configuration options for `webpkgserver`.

> 3. Open `webpkgserver.toml` with an editor of your choice and make the following

Reviewer 2: You might want to mention this optional TOML config:  KeepNonSXGPreloads (default = false).
  # Include preload link headers that lack the corresponding allowed-alt-sxg
  # with a valid header-integrity.

The TOML describes what it's for.

Reviewer 3: If you mention this option, you might want to disclaim that doing so could make the SXG ineligible for the Google SXG cache per [these requirements](https://github.com/google/webpackager/blob/master/docs/cache_requirements.md).

    ```

    `gen-certurl` is a tool that converts certificates to the certificate format
    used by signed exchanges.

2. Verify that `gen-certurl` has been installed correctly.

    ```shell
    gen-certurl --help
    ```

    This command should return information about its usage.

#### Instructions

1. Follow the steps 1 through 3 of [Generate a self-signed
   certificate](/signed-exchanges-webpackager/#generate-a-self-signed-certificate).
2. Convert `cert.pem` to the `application/cert-chain+cbor` format. 

    ```shell
    gen-certurl -pem cert.pem -ocsp &lt;(echo ocsp) > cert.cbor
    ```

    Certificates come in many formats. `cert.pem` is in the [PEM
    format](https://en.wikipedia.org/wiki/Privacy-Enhanced_Mail). Certificates for
    signed exchanges must be in the `application/cert-chain+cbor` format.

    When using a self-signed certificate, `webpkgserver` automatically converts
    the certificate indicated by the `PEMFile` option in `webpkgserver.toml` to
    the `application/cert-chain+cbor` format. However, if you are using a
    `CanSignHttpExchanges` certificate, you must generate the CBOR-encoded

>     certificate yourself.  

Reviewer 3: I think this section is incorrect and should be removed. Web Packager Server will generate cert-chain+cbor even for CanSignHttpExchanges certificates.

(There is a [Web Packager CLI](https://github.com/google/webpackager) that requires use of gen-certurl, but that's not talked about in this doc, so I think safe to omit.)

### Setup up Web Packager for production use

1. Follow steps 1 and 2 of [Setup the Web Packager Server for
   testing](/signed-exchanges-webpackager/#setup-the-web-packager-server-for-testing).

2. Open `webpkgserver.toml` with the editor of your choice and make the
   following changes:

>     *   Change the line `PEMFile = cert.pem` to `PEMFile = cert.cbor`.

Reviewer 3: This line should be changed. PEMFile should point to a PEM file that contains the concatenation of your site's SXG certificate and your CA's certificate.

### Upload the `CanSignHttpExchanges` certificate

1.  Follow steps 1 through 4 of [Upload the self-signed
    certificate](/signed-exchanges-webpackager/#upload-the-self-signed-certificate) to upload `cert.cbor`. 

3. Adjust your server config to serve `cert.cbor` using the `Content-Type:
   application/cert-chain+cbor` response header.

    If this header is not set, you will see the following error when you inspect the
    SXG in DevTools: `Content type of cert-url must be application/cert-chain+cbor.
    Actual content type: text/html`.

4. Restart `webpkgserver`.

    ```shell
    webpkgserver

>     ```

Reviewer 2: As an optional step, you can also talk about how webpkgserver can auto-renew the SXG certs every 90 days if they have a DigiCert ACME account setup.  Relevant config items can be found here: https://github.com/google/webpackager/blob/master/cmd/webpkgserver/webpkgserver.example.toml#L228

>     ```

Reviewer 3: Likewise, this section can be omitted as it only applies to the Web Packager CLI, not the server.
